### DISCUSSION TITLE
Reading from "node:util" is not handled by plugins (Unhandled scheme)

### DISCUSSION THREAD

**Discussion by ShubhamMatoliya** [vercel/next.js#47684] [Help] [answered]
### Summary

while importing a function in middleware.js I am facing this issue. If i don't import it everything works properly. That function basically runs query on database.
**************************************************************************************
error - node:util
Module build failed: UnhandledSchemeError: Reading from "node:util" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.
Import trace for requested module:
node:util
./node_modules/sequelize/lib/dialects/db2/query.js
./node_modules/sequelize/lib/dialects/db2/index.js
./node_modules/sequelize/lib/sequelize.js
./node_modules/sequelize/lib/index.js
./node_modules/sequelize/lib/index.mjs
./models/index.js
./services/websites.js

### Additional information

THIS IS middleware.js FILE IN ROOT DIRECTORY

import { NextResponse } from 'next/server';
// import { retrieveWebsite } from './services/websites';
// importing this function is creating error

export function middleware(request) {
  const {
    nextUrl: { search },
  } = request;
  const urlSearchParams = new URLSearchParams(search);
  //here query = req.query
  const query = Object.fromEntries(urlSearchParams.entries());
  console.log(query);
  return NextResponse.next();
}

**Comment by icyJoseph** [comment]
Could the problem be that the Edge is not a Node environment? https://nextjs.org/docs/api-reference/edge-runtime

It has a different Webpack bundling strategy as well, and various Node APIs won't work at all. 

In particular, https://nodejs.org/api/esm.html#node-imports, this syntax looks like it only applies to Node environments. 

> node: URLs are supported as an alternative means to load Node.js builtin modules. This URL scheme allows for builtin modules to be referenced by valid absolute URL strings.

But the Edge is built on top of V8, rather than being a full Node.js env:

> Edge Functions use Vercel's [Edge Runtime](https://edge-runtime.vercel.sh/packages/runtime), which is built on the same high-performance [V8 JavaScript and WebAssembly engine](https://developers.google.com/apps-script/guides/v8-runtime) used by the Chrome browser.

**Reply by ShubhamMatoliya** [reply]
what could be the solution to get rid of it ? How to fix this?

**Reply by icyJoseph** [reply]
What do you need from util? The idea would be to make your function environment agnostic, but that's not always possible...

**Reply by ShubhamMatoliya** [reply]
utils is used by sequelize package in node modules

**Reply by icyJoseph** [reply]
And this is where most get a bit frustrated, because it just won't work, it is not a Node environment.

I found this in the sequelize GitHub: https://github.com/sequelize/sequelize/issues/14654#issuecomment-1159406912

> That's correct, Sequelize is not designed to run outside of a Node env (for now, there are requests for Deno, react-native, etc).

- https://github.com/sequelize/sequelize/issues/12836

It's not just Edge as you can see, they cannot run on Deno either, at least at the time those comments were written.

**Reply by ShubhamMatoliya** [reply]
Sequelize needs an update.

**Reply by ShubhamMatoliya** [reply]
is there any alternative way or any npm package to fix this problem ?

**Reply by icyJoseph** [reply]
Not sure, but maybe this is a good starting point: https://www.nexxel.dev/blog/typesafe-database

**Reply by AlbertoDente3** [reply]
Same issue: but also with fs,buffer,http,net and https. It shows up after upgrade to webpack 5 from 3.

**Reply by dacmarcell** [reply]
Some update about this?

**Comment by leerob** [comment, marked as answer]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
